[FLINK-14208][python] Optimize Python UDFs with parameters of constant values

HuangXingBo opened the pull request:
## What is the purpose of the change

This pull request optimize Python UDFS with parameters of constant values.The constant parameters are not needed to be transferred between the Java operator and the Python worker.


## Brief change log

  - *Adds Constant message in flink-fn-execution.proto*
  - *Adds translate rexLiteral object to python object logic*

## Verifying this change
This change added tests and can be verified as follows:

  - *Added test test_udf_with_constant_param*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

HuangXingBo commented:
@flinkbot run travis

dianfu commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonPythonCalc.scala:
```diff
@@ -54,6 +57,10 @@ trait CommonPythonCalc {
 
+        case literal: RexLiteral =>
+          inputs.append(
+            rexLiteralToPythonObject.invoke(null, literal.getValue3, literal.getTypeName))
```
Should convert `literal.getValue3` to the actual type. For example, for bigint type, the type of `literal.getValue3` is `BigDecimal`, which is not correct. You can refer to `CodeGenerator#visitLiteral` for more information on how to do that.

dianfu commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonPythonCalc.scala:
```diff
@@ -27,18 +28,20 @@ import scala.collection.mutable
-        "It will be supported in FLINK-14208")
-    }
+    val udfInputOffsets = inputNodes.toArray
+      .map(_._1).filter(_.isInstanceOf[RexInputRef])
```
Move `.filter(xxx)` to a new line?

dianfu commented on flink-python/src/main/java/org/apache/flink/api/common/python/PythonBridgeUtils.java:
```diff
@@ -70,6 +74,50 @@
 		return unpickledData;
 	}
 
+	public static byte[] rexLiteralToPythonObject(Object o, SqlTypeName typeName) {
```
literalToPythonObject?

dianfu commented on flink-python/src/main/java/org/apache/flink/api/common/python/PythonBridgeUtils.java:
```diff
@@ -36,6 +38,8 @@
  */
 public final class PythonBridgeUtils {
 
+	private static Pickler pickler = new Pickler();
```
Make it an local variable of rexLiteralToPythonObject as I think there is no need to reuse it.

dianfu commented on flink-python/pyflink/fn_execution/operations.py:
```diff
@@ -79,6 +82,34 @@ def get(self, value):
+        elif j_type == '\x03' or j_type == 3:
+            self._constant_value = relativedelta(months=pickled_data)
+        elif j_type == '\x04' or j_type == 4:
+            self._constant_value = datetime.timedelta(seconds=pickled_data)
```
throw an meaningful exception if non of the above conditions hold?

dianfu commented on flink-python/pyflink/proto/flink-fn-execution.proto:
```diff
@@ -26,10 +26,12 @@ option java_outer_classname = "FlinkFnApi";
 // User-defined function definition. It supports chaining functions, that's, the execution
 // result of one user-defined function as the input of another user-defined function.
 message UserDefinedFunction {
+
```
unnecessary change

dianfu commented on flink-python/pyflink/table/tests/test_udf.py:
```diff
@@ -96,6 +96,89 @@ def test_udf_in_join_condition(self):
         actual = source_sink_utils.results()
         self.assert_equals(actual, ["2,Hi,2,Flink"])
 
+    def test_udf_with_constant_param(self):
```
1. test_udf_with_constant_param -> test_udf_with_constant_params
2. move it after test_udf_in_join_condition_2?

dianfu commented on flink-python/pyflink/table/tests/test_udf.py:
```diff
@@ -96,6 +96,89 @@ def test_udf_in_join_condition(self):
         self.assert_equals(actual, ["2,Hi,2,Flink"])
 
+    def test_udf_with_constant_param(self):
+        def func_constant_param(p, null_param, tinyint_param, smallint_param, int_param,
```
func_constant_param -> udf_with_constant_params

dianfu commented on flink-python/pyflink/table/tests/test_udf.py:
```diff
@@ -96,6 +96,89 @@ def test_udf_in_join_condition(self):
+
+            p += len(str_param)
+
+            if date_param == datetime.date(year=2014, month=9, day=13):
```
Same as above, throw an exception if the value is not expected.

HuangXingBo commented:
@dianfu  Thanks for your review! I have addressed the comment in the latest commit.

hequn8128 commented:
@HuangXingBo I find that there are problems for supporting the time Interval type in Python due to the equals method of `TimeIntervalTypeInfo`. The equals method use `==` to compare the two serializers. 

This is ok for java because the `INTERVAL_MILLIS` in `TimeIntervalTypeInfo` is a static variable.  However for python, we serialize the `INTERVAL_MILLIS` type in `PythonFunctionCodeGenerator` using `EncodingUtils.encodeObjectToString` and deserialize it later. This makes the two objects can not compare with `==` anymore.

I'm not sure if we can change the compare method in `TimeIntervalTypeInfo`, i.e., change `==` to `equals`. But definitely, it's not a good idea to solve the problem in this PR. We can discuss the problem of the equals method in another jira and support time Interval type later.

What do you think?

Best, Hequn

HuangXingBo commented:
> @HuangXingBo I find that there are problems for supporting the time Interval type in Python due to the equals method of `TimeIntervalTypeInfo`. The equals method use `==` to compare the two serializers.
> 
> This is ok for java because the `INTERVAL_MILLIS` in `TimeIntervalTypeInfo` is a static variable. However for python, we serialize the `INTERVAL_MILLIS` type in `PythonFunctionCodeGenerator` using `EncodingUtils.encodeObjectToString` and deserialize it later. This makes the two objects can not compare with `==` anymore.
> 
> I'm not sure if we can change the compare method in `TimeIntervalTypeInfo`, i.e., change `==` to `equals`. But definitely, it's not a good idea to solve the problem in this PR. We can discuss the problem of the equals method in another jira and support time Interval type later.
> 
> What do you think?
> 
> Best, Hequn

@hequn8128 LGTM,  we can support interval type in another pr later. And I have addressed the comment in the latest commit.

hequn8128 commented on flink-python/pyflink/table/tests/test_udf.py:
```diff
@@ -112,6 +112,103 @@ def test_udf_in_join_condition_2(self):
+                                     bigint_param, decimal_param, float_param, double_param,
+                                     boolean_param, str_param,
+                                     date_param, time_param, timestamp_param):
+            # decide two float whether equals
```
decide whether two floats are equal

hequn8128 commented:
I have created a new jira([FLINK-14383](https://issues.apache.org/jira/browse/FLINK-14383)) for the time interval types.

hequn8128 commented:
@HuangXingBo Thanks a lot for the update. LGTM, will merge this when Travis passed.
